### PR TITLE
Fix remaining edb.pgsql.compiler typing errors

### PR DESCRIPTION
**Pull request by elprans**
This makes mypy (strictly) happy with `edb.pgsql.compiler`.

### CODE REVIEW

File: edb/common/compiler.py [hunk 1]
```diff
@@ -76,17 +99,17 @@ def _current(self):
...
-    def nextval(self, name='default'):
+    def nextval(self, name: str = 'default') -> int:
         self.counts[name] += 1
         return self.counts[name]
 
 
-class AliasGenerator(Counter):
-    def get(self, hint=None):
+class AliasGenerator(SimpleCounter):
+    def get(self, hint: Optional[str]=None) -> str:
```
Comment (ambv) [inline edb/common/compiler.py]: If an empty string can be a sentinel, it's generally preferred. In this case could 'v' be the default?

File: edb/ir/scopetree.py [hunk 2]
```diff
@@ -585,13 +585,15 @@ def find_unfenced(self, path_id: pathid.PathId) \
...
+        result: Optional[ScopeTreeNode] = None
+
         for node in self.descendants:
             if node.unique_id == unique_id:
-                return node
+                result = node
+                break
 
-        return None
+        return result
```
Comment (ambv) [inline edb/ir/scopetree.py]: Did you have to make those changes to make Mypy happy?

File: edb/pgsql/common.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ def get_module_backend_name(module, prefix='edgedb_'):
     return edgedb_name_to_pg_name(f'{prefix}{module}', len(prefix))
 
 
-def edgedb_name_to_pg_name(name, prefix_length=0):
+def edgedb_name_to_pg_name(name, prefix_length=0) -> str:
```
Comment (ambv) [inline edb/pgsql/common.py]: `name` and `prefix_length` look like easy targets to type, too?

File: edb/pgsql/compiler/dml.py [hunk 4]
```diff
@@ -440,7 +440,7 @@ def compile_insert_shape_element(
             # anyway.
             insvalctx.volatility_ref = context.NO_VOLATILITY
 
-        insvalctx.path_scope[ir_stmt.subject.path_id] = insert_stmt
+        # insvalctx.path_scope[ir_stmt.subject.path_id] = insert_stmt
```
Comment (ambv) [inline edb/pgsql/compiler/dml.py]: Is this supposed to stay commented out?

File: edb/pgsql/compiler/dml.py [hunk 5]
```diff
@@ -415,7 +415,7 @@ def process_insert_body(
                     shape_el, None, ctx=scopectx)
                 prop_elements.append(tuple_el)
 
-        valtuple = pgast.TupleVar(elements=prop_elements, named=True)
+        valtuple = pgast.TupleVarBase(elements=prop_elements, named=True)
```
Comment (ambv) [inline edb/pgsql/compiler/dml.py]: Why the change? Esp. that typically I understand a class called "*Base" to be a building block, not something to be directly instantiated.

File: edb/pgsql/compiler/pathctx.py [hunk 6]
```diff
@@ -124,6 +133,7 @@ def get_path_var(
             src_path_id = path_id
         else:
             src_path_id = path_id.src_path()
+            assert src_path_id is not None
```
Comment (ambv) [inline edb/pgsql/compiler/pathctx.py]: This is subtle and I can't see. I looked in ir/pathid.py and in the src_path() method but I don't quite get it. Why do we know this will never be None?

File: edb/pgsql/compiler/pathctx.py [hunk 7]
```diff
@@ -265,19 +280,6 @@ def get_path_var(
...
-        else:
-            outer_path_id = drilldown_path_id
-
-        path_id_map = {
-            outer_path_id: next(iter(source_rel.path_scope))
-        }
-
-        drilldown_path_id = map_path_id(
-            drilldown_path_id, path_id_map)
-
```
Comment (ambv) [inline edb/pgsql/compiler/pathctx.py]: Dead code?

File: mypy.ini [hunk 8]
```diff
@@ -16,6 +16,22 @@ warn_unused_configs = True
 follow_imports = True
 ignore_errors = False
 
+[mypy-edb.pgsql.compiler.*]
+follow_imports = True
+ignore_errors = False
+# Equivalent of --strict on the command line:
+disallow_subclassing_any = True
+disallow_any_generics = True
+; disallow_untyped_calls = True
```
Comment (ambv) [inline mypy.ini]: Use # for comments and comment why you're disabling this. In checked.py the only reason is use of @overload which requires the default implementation to be untyped (this is where PEP 484 disagrees with Mypy I guess).

File: edb/ir/scopetree.py (hunk 2 above)
Comment (elprans) [reply edb/ir/scopetree.py]: Yes, otherwise it complained that I was returning `Any`

File: edb/pgsql/compiler/dml.py (hunk 5 above)
Comment (elprans) [reply edb/pgsql/compiler/dml.py]: `TupleVarBase` differs from `TupleVar` in that the former is just a "shape" of a tuple, whereas the latter actually holds the values.  In this case we need the one without the values.  Another approach is to make `TupleElement.val` to be `Optional`, but that opens a much bigger can of worms.

File: edb/ir/scopetree.py (hunk 2 above)
Comment (ambv) [reply edb/ir/scopetree.py]: That looks like a bug. It probably can't unwind the double generator right?

File: edb/pgsql/compiler/pathctx.py (hunk 6 above)
Comment (elprans) [reply edb/pgsql/compiler/pathctx.py]: It is correlated with `path_id.rptr()`, which we check above.  If `path_id.rptr()` is not `None`, then `src_path()` would not be.  Both are essentially `[:-1]` slices on the path id.

File: edb/pgsql/compiler/pathctx.py (hunk 7 above)
Comment (elprans) [reply edb/pgsql/compiler/pathctx.py]: Dead code.

File: mypy.ini (hunk 8 above)
Comment (elprans) [reply mypy.ini]: VSCode did the `;` thing, not me :-)  `disallow_untyped_calls` is unset because the compiler calls out to lots of untyped code in `edb.pgsql`, which is beyond the scope here.

File: edb/common/compiler.py [hunk 9]
```diff
@@ -19,55 +19,78 @@
...
+
+
 class ContextLevel:
     def on_pop(self, prevlevel):
         pass
 
-    def new(self, mode=None):
+    def new(
+        self: ContextLevel_T,
+        mode=None,
```
[the remark this answers is not in the document]
Comment (elprans) [reply edb/common/compiler.py]: Now is.

File: edb/common/compiler.py [hunk 10]
```diff
@@ -19,55 +19,78 @@
...
         return self._stack.new(mode, self)
 
 
-class CompilerContextManager:
-    def __init__(self, context, mode, prevlevel):
+class CompilerContextManager(Generic[ContextLevel_T]):
+    def __init__(
+        self,
+        context: CompilerContext[ContextLevel_T],
+        mode: Any,
```
[the remark this answers is not in the document]
Comment (elprans) [reply edb/common/compiler.py]: Yeah, depends on actual context implementation, usually some enum.

File: edb/common/compiler.py (hunk 1 above)
Comment (elprans) [reply edb/common/compiler.py]: Fixed.

File: edb/ir/scopetree.py (hunk 2 above)
Comment (elprans) [reply edb/ir/scopetree.py]: Yeah, this was probably some intermediate issue.  It's not reproducing now, so I reverted back to the original formulation.

File: edb/pgsql/common.py (hunk 3 above)
Comment (elprans) [reply edb/pgsql/common.py]: Yes.

File: edb/pgsql/compiler/dml.py (hunk 4 above)
Comment (elprans) [reply edb/pgsql/compiler/dml.py]: It's supposed to be deleted, which I've now done.
